Well-Behaved (Co)algebraic Semantics of Regular Expressions in Dafny (#23)
dafny-lang/blog: _posts/2024-01-10-semantics-of-regular-expressions.markdown

## Denotational Semantics

In this section, we define regular expressions and formal languages, introduce the concept of bisimilarity, formalise the *denotational* semantics of regular expressions as a function from regular expressions to formal languages, and prove that the latter is an algebra homomorphism.

### Regular Expressions as Datatype

We define the set of regular expressions parametric in an alphabet  `A` as an inductive [`datatype`](https://dafny.org/latest/DafnyRef/DafnyRef#sec-inductive-datatypes):

```
datatype Exp<A> = Zero | One | Char(A) | Plus(Exp, Exp) | Comp(Exp, Exp) | Star(Exp)
```

> The definition captures that a regular expression is either a primitive character `Char(a)`,  a non-deterministic choice between two regular expressions `Plus(e1, e2)`, a sequential iteration of two regular expressions `Comp(e1, e2)`,  a finite number of self-iterations `Star(e)`, or one of the constants `Zero` (the unit of `Plus`) and `One` (the unit of `Comp`). On a more high-level, above defines  `Exp<A>` as the *smallest* algebraic structures that is equipped with two constants, contains all elements of type `A`, and is closed under two binary and one unary operation.

Reviewer: I'm not sure what "sequential iteration" is supposed to mean. Why not just call it "concatenation"? Similarly, can we say something like "Zero" is the empty set and "One" the empty string, just to give more intuition to the readers?

Author: As suggested by Aaron, I'll go with sequential _composition_.

### Formal Languages as Codatatype

We define the set of formal languages parametric in an alphabet `A` as an coinductive [`codatatype`](https://dafny.org/latest/DafnyRef/DafnyRef#sec-coinductive-datatypes):

```
codatatype Lang<!A> = Alpha(eps: bool, delta: A -> Lang<A>)
```

> To some, this choice might seem odd at first sight. If you are familiar with the topic, you likely have expected the set of formal languages to be defined more concretely as the set of all sets of finite sequences (sometimes called *words*),  `iset<seq<A>>`. Rest assured, we agree — up to an appropriate notion of equality! Where as you characterise languages intrinsically, we treat them extrinsically, in terms of their universal property: [it is well known](https://ir.cwi.nl/pub/28550/rutten.pdf) that `iset<seq<A>>` forms the *greatest* [coalgebraic](https://en.wikipedia.org/wiki/F-coalgebra) structure (think of a deterministic automaton without initial state) `S` that is equipped with functions `eps: S -> bool` and `delta: S → (A → S)`. Indeed, for any set `U` of finite sequences, we can verify whether `U` contains the empty sequence, `U.eps == ([] in U)`, and for any `a: A`, we can transition to the set  `U.delta(a) == (iset s | [a] + s in U)`. Here, we choose the more abstract perspective on formal languages at it hides irrelevant specifics and thus allows us to write more elegant proofs.

Reviewer 2: ```suggestion
To some, this choice might seem odd at first sight. If you are familiar with the topic, you likely have expected the set of formal languages to be defined more concretely as the set of all sets of finite sequences (sometimes called *words*),  `iset<seq<A>>`. Rest assured, we agree — up to an appropriate notion of equality! Whereas you characterise languages intrinsically, we treat them extrinsically, in terms of their universal property: [it is well known](https://ir.cwi.nl/pub/28550/rutten.pdf) that `iset<seq<A>>` forms the *greatest* [coalgebraic](https://en.wikipedia.org/wiki/F-coalgebra) structure (think of a deterministic automaton without initial state) `S` that is equipped with functions `eps: S -> bool` and `delta: S → (A → S)`. Indeed, for any set `U` of finite sequences, we can verify whether `U` contains the empty sequence, `U.eps == ([] in U)`, and for any `a: A`, we can transition to the set  `U.delta(a) == (iset s | [a] + s in U)`. Here, we choose the more abstract perspective on formal languages at it hides irrelevant specifics and thus allows us to write more elegant proofs.
```

### An Algebra of Formal Languages

If you think of formal languages as the set of all sets of finite sequences, you will soon realise that languages admit quite a bit of algebraic structure. For example, there exist two languages of distinct importance (can you already guess which ones?), and one can obtain a new language by taking e.g. the union of two languages. In fact, if you think about it for a bit longer, you’ll realise that formal languages admit exactly the same [type of algebraic structure](https://en.wikipedia.org/wiki/F-algebra) as the one you’ve encountered when we defined regular expressions!

> First, there exists the empty language `Zero()` that contains no words at all. Under above view, we find `Zero().eps == false` and `Zero().delta(a) == Zero`, since the empty set does not contain the empty sequence, and the derivative `iset s | [a] + s in iset{}` with respect to any `a: A` yields again the empty set, respectively. We thus define:

Reviewer: Zero -> Zero()

```
ghost function Zero<A>(): Lang {
  Alpha(false, (a: A) => Zero())
}
```

Using similar reasoning, we additionally derive the following definitions. In order, we formalise i) the language `One()` that contains only the empty sequence; ii) for any `a: A` the language `Singleton(a)` that consists of only the word `[a]`; iii) the language `Plus(L1, L2)` which consists of the union of the languages `L1` and `L2`; iv) the language `Comp(L1, L2)` that consists of all possible concatenation of words in `L1` and `L2`; and v) the language `Star(L)` that consists of all finite compositions of `L` with itself. Our definitions match what is well-known as *[Brzozowski derivatives](https://en.wikipedia.org/wiki/Brzozowski_derivative)*. 

```
ghost function One<A>(): Lang {
  Alpha(true, (a: A) => Zero())
}

ghost function Singleton<A>(a: A): Lang {
  Alpha(false, (b: A) => if a == b then One() else Zero())
}

> ghost function {:abstemious} Plus<A>(L1: Lang, L2: Lang): Lang {

Reviewer 2: It might be worth introducing the idea of abstemious functions here?

```
ghost function Eps<A>(e: Exp): bool {
  match e 
    case Zero => false
    case One => true
    case Char(a) => false
    case Plus(e1, e2) => Eps(e1) || Eps(e2)
    case Comp(e1, e2) => Eps(e1) && Eps(e2)
    case Star(e1) => true
}

ghost function Delta<A(!new)>(e: Exp): A -> Exp {
  (a: A) => 
    match e 
      case Zero => Zero
      case One => Zero
      case Char(b) => if a == b then One else Zero
      case Plus(e1, e2) => Plus(Delta(e1)(a), Delta(e2)(a))
      case Comp(e1, e2) => Plus(Comp(Delta(e1)(a), e2), Comp(if Eps(e1) then One else Zero, Delta(e2)(a)))
      case Star(e1) => Comp(Delta(e1)(a), Star(e1))
}
```

### Operational Semantics as Induced Morphism

The operational semantics of regular expressions can now be defined via coinduction, as a function `Operational: Exp -> Lang`, by making use of the coalgebraic structure on expressions we have just defined:

```
function Operational<A(==)>(e: Exp): Lang {
  Alpha(Eps(e), (a: A) => Operational(Delta(e)(a)))
}
```

### Operational Semantics as Coalgebra Homomorphism

> In [Denotational Semantics as Algebra Homomorphism](#denotational-semantics-as-algebra-homomorphism) we defined algebra homomorphisms as functions `f: Exp -> Lang` that commute with the algebraic structures of regular expressions and formal languages, respectively. Analogously, let us now call a function `f` of the same type a *coalgebra homomorphism*, if it commutes with the *coalgebraic* structures of regular expressions and formal languages, respectively:

Reviewer: The link is broken in the Github rendering. Not sure how it'll render on the blog.

Author: The link works

```
lemma OperationalIsCoalgebraHomomorphism<A(!new)>()
  ensures IsCoalgebraHomomorphism<A>(Operational)
{
  forall e, a ensures Bisimilar<A>(Operational(e).delta(a), Operational(Delta(e)(a))) {
    BisimilarityIsReflexive(Operational(e).delta(a));
  }
}
```

## Well-Behaved Semantics

So far, we have seen two dual approaches for assigning a formal language semantics to regular expressions:

* `Denotational`: an algebra homomorphism obtained via induction
* `Operational`: a coalgebra homomorphism obtained via coinduction

Next, we show that the denotational and operational semantics of regular expressions are *well-behaved*: they constitute two sides of the same coin. First, we show that `Denotational` is also a coalgebra homomorphism, and that coalgebra homomorphisms are unique up to bisimulation. We then deduce from the former that `Denotational` and `Operational` coincide pointwise, up to bisimulation. Finally, we show that `Operational` is also an algebra homomorphism.

### Denotational Semantics as Coalgebra Homomorphism

> In this section, we establish that `Denotational` does not only commute with the algebraic structures of regular expressions and formal languages, but also with their coalgebraic structures:

Reviewer 2: ```suggestion
In this section, we establish that `Denotational` not only commutes with the algebraic structures of regular expressions and formal languages, but also with their coalgebraic structures:
```

  UniqueCoalgebraHomomorphism<A>(Operational, Denotational, e);
}
```

### Operational Semantics as Algebra Homomorphism

As a little extra, for the sake of symmetry, let us also prove that `Operational` is an algebra homomorphism. (We already know that it is a coalgebra homomorphism, and that `Denotational` is both an algebra and coalgebra homomorphism.)

```
lemma OperationalIsAlgebraHomomorphism<A(!new)>()
  ensures IsAlgebraHomomorphism<A>(Operational)
{...}
```

The idea of the proof is to take advantage of `Denotational` being an algebra homomorphism, by translating its properties to `Operational` via the lemma in [Denotational and Operational Semantics Are Bisimilar](#denotational-and-operational-semantics-are-bisimilar).  The relevant new statements capture that bisimilarity is symmetric and a congruence with respect to the `Star` operation:

```
greatest lemma BisimilarityIsSymmetric<A(!new)>[nat](L1: Lang, L2: Lang)
  ensures Bisimilar(L1, L2) ==> Bisimilar(L2, L1)
  ensures Bisimilar(L1, L2) <== Bisimilar(L2, L1)
{}

lemma StarCongruence<A(!new)>(L1: Lang, L2: Lang)
  requires Bisimilar(L1, L2)
  ensures Bisimilar(Star(L1), Star(L2))
{...}
```

## Conclusion

> We have used Dafny’s built-in inductive and coinductive reasoning capabilities to define two language semantics for regular expressions: denotational and operational semantics. Through a number of dualities — construction and deconstruction, algebras and coalgebras, and congruence and bisimilarity — we have proven the semantics to be two sides of the same coin. The blogpost is inspired by research in the field of [Coalgebra](https://en.wikipedia.org/wiki/F-coalgebra), which was pioneered by [Rutten](https://pdf.sciencedirectassets.com/271538/1-s2.0-S0304397500X01466/1-s2.0-S0304397500000566/main.pdf), [Gumm](https://www.researchgate.net/publication/2614339_Elements_Of_The_General_Theory_Of_Coalgebras), and others. The concept of well-behaved semantics goes back to [Turi and Plotkin](https://homepages.inf.ed.ac.uk/gdp/publications/Math_Op_Sem.pdf) and was adapted by [Jacobs](https://link.springer.com/chapter/10.1007/11780274_20) to the case of regular expressions. We heavily used automata theoretic constructions from the 1960s, originally investigated by [Brzozowski](https://dl.acm.org/doi/10.1145/321239.321249) (a more modern presentation can be found e.g. [here](https://alexandrasilva.org/files/thesis.pdf)). Due to space constraints, we had to exclude the details of most of the proofs. To dive deep, please take a look at the full Dafny source code, which is available [here](../../../../assets/src/semantics-of-regular-expressions/Languages.dfy), [here](../../../../assets/src/semantics-of-regular-expressions/Semantics.dfy), and [here](../../../../assets/src/semantics-of-regular-expressions/Expressions.dfy).

Reviewer 2: I don't know that it really makes sense to say "due to space constraints" in a blog post. Maybe say something about focusing on the most important intuitive aspects of the proof, instead?

### Regular Expressions as Datatype

We define the set of regular expressions parametric in an alphabet  `A` as an inductive [`datatype`](https://dafny.org/latest/DafnyRef/DafnyRef#sec-inductive-datatypes):

```
datatype Exp<A> = Zero | One | Char(A) | Plus(Exp, Exp) | 
                  Comp(Exp, Exp) | Star(Exp)
```

The definition captures that a regular expression is either a primitive character `Char(a)`,  a non-deterministic choice between two regular expressions `Plus(e1, e2)`, a sequential composition of two regular expressions `Comp(e1, e2)`,  a finite number of self-iterations `Star(e)`, or one of the constants `Zero` (the unit of `Plus`) and `One` (the unit of `Comp`). At a higher level, the above defines  `Exp<A>` as the *smallest* algebraic structure that is equipped with two constants, contains all elements of type `A`, and is closed under two binary operations and one unary operation.

### Formal Languages as Codatatype

We define the set of formal languages parametric in an alphabet `A` as an coinductive [`codatatype`](https://dafny.org/latest/DafnyRef/DafnyRef#sec-coinductive-datatypes):

```

> codatatype Lang<!A> = Alpha(eps: bool, delta: A -> Lang<A>)

Author: I added a paragraph.

### Regular Expressions as Datatype

We define the set of regular expressions parametric in an alphabet  `A` as an inductive [`datatype`](https://dafny.org/latest/DafnyRef/DafnyRef#sec-inductive-datatypes):

```
datatype Exp<A> = Zero | One | Char(A) | Plus(Exp, Exp) | 
                  Comp(Exp, Exp) | Star(Exp)
```

Note that above, and later, we make use of Dafny's [type parameter completion](https://leino.science/papers/krml270.html).

The definition captures that a regular expression is either a primitive character `Char(a)`,  a non-deterministic choice between two regular expressions `Plus(e1, e2)`, a sequential composition of two regular expressions `Comp(e1, e2)`,  a finite number of self-iterations `Star(e)`, or one of the constants `Zero` (the unit of `Plus`) and `One` (the unit of `Comp`). At a higher level, the above defines  `Exp<A>` as the *smallest* algebraic structure that is equipped with two constants, contains all elements of type `A`, and is closed under two binary operations and one unary operation.

### Formal Languages as Codatatype

> We define the set of formal languages parametric in an alphabet `A` as an coinductive [`codatatype`](https://dafny.org/latest/DafnyRef/DafnyRef#sec-coinductive-datatypes):

Reviewer 2: ```suggestion
We define the set of formal languages parametric in an alphabet `A` as a coinductive [`codatatype`](https://dafny.org/latest/DafnyRef/DafnyRef#sec-coinductive-datatypes):
```
